samples/join-a-team-using-qr-code/csharp/JoinTeamByQR/TokenExchangeHelper.cs

```
@@ -0,0 +1,162 @@
+
+            // This example utilizes Bot Framework IStorage's ETag implementation for token exchange activity deduplication.
+
+            // Create a StoreItem with Etag of the unique 'signin/tokenExchange' request
+            var storeItem = new TokenStoreItem
+            {
+                ETag = (turnContext.Activity.Value as JObject).Value<string>("id")
+            };
+
+            var storeItems = new Dictionary<string, object> { { TokenStoreItem.GetStorageKey(turnContext), storeItem } };
+            try
+            {
+                // Writing the IStoreItem with ETag of unique id will succeed only once
+                await _storage.WriteAsync(storeItems, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+                // Memory storage throws a generic exception with a Message of 'Etag conflict. [other error info]'
+                when (ex.Message.StartsWith("Etag conflict"))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private async Task<bool> ExchangedTokenAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            TokenResponse tokenExchangeResponse = null;
+            var tokenExchangeRequest = ((JObject)turnContext.Activity.Value)?.ToObject<TokenExchangeInvokeRequest>();
+
+            try
+            {
+                tokenExchangeResponse = await (turnContext.Adapter as IExtendedUserTokenProvider).ExchangeTokenAsync(
+                    turnContext,
+                    _oAuthConnectionName,
+                    turnContext.Activity.From.Id,
+                    new TokenExchangeRequest { Token = tokenExchangeRequest.Token },
+                    cancellationToken).ConfigureAwait(false);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types (ignoring, see comment below)
```

Why do we need these?

---

Those are included to ignore exceptions. The format has been used from old samples so implemented the same here